tests/lib/linter/source-code-traverser.js

```diff
@@ -0,0 +1,387 @@
+				emitter.on(selector, () => {}),
+			);
+			traverser = SourceCodeTraverser.getInstance(MOCK_LANGUAGE);
+		});
+
+		it("should generate events for AST nodes.", () => {
+			const dummyNode = { type: "Foo", value: 1 };
+			const sourceCode = createMockSourceCode(dummyNode);
+
+			traverser.traverseSync(sourceCode, emitter);
+
+			assert(emitter.emit.calledTwice);
+			assert(emitter.emit.firstCall.calledWith("Foo", dummyNode));
+			assert(emitter.emit.secondCall.calledWith("Foo:exit", dummyNode));
+		});
+
+		it("should generate events for nested AST nodes", () => {
+			const dummyNode = {
+				type: "Foo",
+				value: 1,
+				child: { type: "Bar", value: 2 },
+			};
+
+			const sourceCode = createMockSourceCode(dummyNode);
+
+			traverser.traverseSync(sourceCode, emitter);
+
+			assert(emitter.emit.callCount === 4);
+			assert(
+				emitter.emit.firstCall.calledWith("Foo", dummyNode),
+				"First call was wrong",
+			);
+			assert(
+				emitter.emit.secondCall.calledWith("Bar", dummyNode.child),
+				"Second call was wrong",
+			);
+			assert(
+				emitter.emit.thirdCall.calledWith("Foo > Bar", dummyNode.child),
+				"Third call was wrong",
+			);
+			assert(
+				emitter.emit.lastCall.calledWith("Foo:exit", dummyNode),
+				"Last call was wrong",
+			);
+		});
+
+		it("should handle call steps in traverse", () => {
+			const dummyNode = { type: "Foo", value: 1 };
+			const sourceCode = {
+				ast: dummyNode,
+				visitorKeys: vk.KEYS,
+				*traverse() {
+					yield {
+						kind: STEP_KIND_VISIT,
+						target: dummyNode,
+						phase: 1,
+					};
+					yield {
+						kind: STEP_KIND_CALL,
+						target: "customEvent",
+						args: [dummyNode, "extra"],
+					};
+					yield {
+						kind: STEP_KIND_VISIT,
+						target: dummyNode,
+						phase: 2,
+					};
+				},
+			};
+
+			traverser.traverseSync(sourceCode, emitter);
+
+			assert(emitter.emit.calledThrice);
+			assert(emitter.emit.firstCall.calledWith("Foo", dummyNode));
+			assert(
+				emitter.emit.secondCall.calledWith(
+					"customEvent",
+					dummyNode,
+					"extra",
+				),
+			);
+			assert(emitter.emit.thirdCall.calledWith("Foo:exit", dummyNode));
+		});
+
+		it("should throw error for invalid step kind", () => {
+			const dummyNode = { type: "Foo", value: 1 };
+			const sourceCode = {
+				ast: dummyNode,
+				visitorKeys: vk.KEYS,
+				*traverse() {
+					yield {
+						kind: 999, // Invalid step kind
+						target: dummyNode,
+					};
+				},
+			};
+
+			assert.throws(
+				() => traverser.traverseSync(sourceCode, emitter),
+				/Invalid traversal step found:/u,
+			);
+		});
+
+		it("should throw error with currentNode property when error occurs during traversal", () => {
+			const dummyNode = { type: "Foo", value: 1 };
+			const emitterWithError = Object.create(createEmitter(), {
+				emit: {
+					value() {
+						throw new Error("Test error");
+					},
+				},
+			});
+
+			["Foo"].forEach(selector =>
+				emitterWithError.on(selector, () => {}),
+			);
```

Comment:
```suggestion
			const emitterWithError = createEmitter();

			emitterWithError.on("Foo", () => {
				throw new Error("Test error");
			});
```

To simplify the test, and I think it's more representative that a listener rather than the emitter throws an error.